File: Day12/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		int north = 0;
		int west = 0;
		int waypointNorth = 1;
		int waypointWest = -10;
		int shipNorth = 0;
		int shipWest = 0;
		int rotation = 1;
		for (String line : Files.readAllLines(input.toPath())) {
			int value = Integer.parseInt(line.substring(1));
			switch (line.charAt(0)) {
			case 'N':
				north += value;
				waypointNorth += value;
				break;
			case 'S':
				north -= value;
				waypointNorth -= value;
				break;
			case 'E':
				west -= value;
				waypointWest -= value;
				break;
			case 'W':
				west += value;
				waypointWest += value;
				break;
			case 'L':
				value /= 90;
				rotation = (rotation - value) % 4;
				while (rotation < 0) {
					rotation += 4;
				}
				int oldWest = waypointWest;
				if (value == 1) {
					waypointWest = waypointNorth;
					waypointNorth = -oldWest;
				} else if (value == 2) {
					waypointWest *= -1;
					waypointNorth *= -1;
				} else if (value == 3) {
					waypointWest = -waypointNorth;
					waypointNorth = oldWest;
				}
				break;
			case 'R':
				value /= 90;
				rotation = (rotation + value) % 4;
				while (rotation < 0) {
					rotation += 4;
				}
				oldWest = waypointWest;
				if (value == 1) {
					waypointWest = -waypointNorth;
					waypointNorth = oldWest;
				} else if (value == 2) {
					waypointWest *= -1;
					waypointNorth *= -1;
				} else if (value == 3) {
					waypointWest = waypointNorth;
					waypointNorth = -oldWest;
				}
				break;
			case 'F':
				if (rotation == 0) {
					north += value;
				} else if (rotation == 1) {
					west -= value;
				} else if (rotation == 2) {
					north -= value;
				} else if (rotation == 3) {
					west += value;
				}

				shipNorth += waypointNorth;
				shipWest += waypointWest;
			}
		}

		System.out.println("The Manhattan distance from the ferrys starting position to its part 1 position is "
				+ (Math.abs(north) + Math.abs(west)) + ".");

		System.out.println("The Manhattan distance from the ferrys starting position to its part 2 position is "
				+ (Math.abs(shipNorth) + Math.abs(shipWest)) + ".");
	}
}

File: Day12/Python/Main.py
import os

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        print(f"trying {input} instead.")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    inputFile = open(input, 'r')
    north = 0
    west = 0
    waypointNorth = 1;
    waypointWest = -10;
    shipNorth = 0;
    shipWest = 0;
    rotation = 1
    for line in inputFile.readlines():
        value = int(line[1:-1])
        if line[0] == 'N':
            north += value
            waypointNorth += value
        elif line[0] == 'S':
            north -= value
            waypointNorth -= value
        elif line[0] == 'E':
            west -= value
            waypointWest -= value
        elif line[0] == 'W':
            west += value
            waypointWest += value
        elif line[0] == 'L':
            value /= 90
            rotation = (rotation - value) % 4
            oldWest = waypointWest
            if value == 1:
                waypointWest = waypointNorth
                waypointNorth = -oldWest
            elif value == 2:
                waypointWest *= -1
                waypointNorth *= -1
            elif value == 3:
                waypointWest = -waypointNorth
                waypointNorth = oldWest
        elif line[0] == 'R':
            value /= 90
            rotation = (rotation + value) % 4
            oldWest = waypointWest
            if value == 1:
                waypointWest = -waypointNorth
                waypointNorth = oldWest
            elif value == 2:
                waypointWest *= -1
                waypointNorth *= -1
            elif value == 3:
                waypointWest = waypointNorth
                waypointNorth = -oldWest
        elif line[0] == 'F':
            if rotation == 0:
                north += value
            elif rotation == 1:
                west -= value
            elif rotation == 2:
                north -= value
            elif rotation == 3:
                west += value

            shipNorth += waypointNorth * value
            shipWest += waypointWest * value

    inputFile.close()
    print(f"The Manhattan distance from the ferrys starting position to its part 1 position is {abs(north) + abs(west)}.")

    print(f"The Manhattan distance from the ferrys starting position to its part 2 position is {abs(shipNorth) + abs(shipWest)}.")

if __name__ == '__main__':
    main()
